proof/crefine/AARCH64/Retype_C.thy

```diff
@@ -5866,31 +5901,62 @@ proof -
-        apply (clarsimp simp: hrs_htd_update ptBits_def objBits_simps
-                              AARCH64_H.createObject_def pageBits_def ptTranslationBits_def
-                              cond_second_eq_seq_ccorres modify_gsUserPages_update
-                        intro!: ccorres_rhs_assoc)
-        apply ((rule ccorres_return_C | simp | wp | vcg
-           | (rule match_ccorres, ctac add:
-                   placeNewDataObject_ccorres[where us=18 and newType=newType, simplified]
-                   gsUserPages_update_ccorres[folded modify_gsUserPages_update])
-           | (rule match_ccorres, csymbr))+)[1]
-       apply (intro conjI)
+    subgoal
+        apply (in_case "HugePageObject")
+        apply (cinit' lift: t_' regionBase_' userSize_' deviceMemory_')
+         apply (simp add: object_type_from_H_def Kernel_C_defs)
+         apply (simp add: ccorres_cond_univ_iff ccorres_cond_empty_iff
+                          asidInvalid_def APIType_capBits_def shiftL_nat objBits_simps
+                          ptBits_def pageBits_def word_sle_def word_sless_def fold_eq_0_to_bool)
+         apply (clarsimp simp: hrs_htd_update ptBits_def objBits_simps
+                               AARCH64_H.createObject_def pageBits_def ptTranslationBits_def
+                               cond_second_eq_seq_ccorres modify_gsUserPages_update
+                         intro!: ccorres_rhs_assoc)
+         apply (rule ccorres_cases[where P=deviceMemory])
+          apply clarsimp
+          apply (rule ccorres_rhs_assoc)+
+          apply (rule ccorres_split_nothrow_novcg)
+              apply (rule placeNewDataObject_dev_ccorres[where us=18 and newType=newType, simplified])
+             apply ceqv
+            apply  (ctac (no_vcg) add: gsUserPages_update_ccorres[folded modify_gsUserPages_update])
+              apply csymbr
+              apply (rule ccorres_return_C; simp)
+             apply wp
+            apply (clarsimp simp: pageBits_def ccap_relation_def APIType_capBits_def
+                        framesize_to_H_def cap_to_H_simps cap_frame_cap_lift
+                        vm_page_size_defs ptTranslationBits_def
+                        canonical_address_and_maskD[unfolded mask_def, simplified]
+                        vmrights_to_H_def mask_def vm_rights_defs c_valid_cap_def cl_valid_cap_def)
+           apply wp
+          apply (simp add: guard_is_UNIV_def)
+         apply clarsimp
+         apply (rule ccorres_rhs_assoc)+
+         apply (rule ccorres_split_nothrow_novcg)
+             apply (rule placeNewDataObject_no_dev_ccorres[where us=18 and newType=newType, simplified])
+            apply ceqv
+           apply  (ctac (no_vcg) add: gsUserPages_update_ccorres[folded modify_gsUserPages_update])
+             apply csymbr+
+             apply (rule ccorres_Guard_Seq)
+             apply (ctac (no_vcg) add: cleanCacheRange_RAM_ccorres)
+              apply csymbr
+              apply (rule ccorres_return_C; simp)
+             apply wp
```

Comment:
out of curiosity: is there any reason why you prefer wp over wpsimp in this proof?

Reply:
If the `simp` is not necessary, I tend to prefer just `wp` because it's the less complex method and adding clarsimp into the mix at the end can mask problems later (although it can also solve simple unexpected new side goals, so it's not a completley clear cut argument -- I think I just like to use the simpler methods when they should suffice)